cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -0,0 +1,294 @@
+      if (!sources.contains(node) && !connectors.contains(node)) {
+        continue;
+      }
+
+      Set<String> accessibleByNode = accessibleFrom(node, Sets.union(connectors, reduceNodes));
+      Set<String> sinksAndReduceNodes = Sets.intersection(accessibleByNode, Sets.union(sinks, reduceNodes));
+      // don't count this node
+      sinksAndReduceNodes = Sets.difference(sinksAndReduceNodes, ImmutableSet.of(node));
+
+      if (sinksAndReduceNodes.size() > 1) {
+        for (String reduceNodeConnector : Sets.intersection(sinksAndReduceNodes, reduceNodes)) {
+          addConnectorInFrontOf(reduceNodeConnector, addedAlready);
+        }
+      }
+    }
+
+    /*
+        Find nodes that have input from multiple sources and add them to the connectors set.
+        We can probably remove this part once we support multiple sources. Even though we don't support
+        multiple sources today, the fact that we support forks means we have to deal with the multi-input case
+        and break it down into separate phases. For example:
+
+                    |---> reduce1 ---|
+          source ---|                |---> sink
+                    |---> reduce2 ---|
+
+        From the previous section, both reduces will get a connector inserted in front:
+
+                    |---> reduce1.connector               reduce1.connector ---> reduce1 ---|
+          source ---|                              =>                                       |---> sink
+                    |---> reduce2.connector               reduce2.connector ---> reduce2 ---|
+
+        Since we don't support multi-input yet, we need to convert that further into 3 phases:
+
+          reduce1.connector ---> reduce1 ---> sink.connector
+                                                                    =>       sink.connector ---> sink
+          reduce2.connector ---> reduce2 ---> sink.connector
+
+        To find these nodes, we traverse the graph in order and keep track of sources that have a path to each node
+        with a map of node -> [ sources that have a path to the node ]
+        if we find that a node is accessible by more than one source, we insert a connector in front of it and
+        reset all sources for that node to its connector
+     */
+    SetMultimap<String, String> nodeSources = HashMultimap.create();
+    for (String source : sources) {
+      nodeSources.put(source, source);
+    }
+    for (String node : getTopologicalOrder()) {
+      if (sinks.contains(node)) {
+        continue;
+      }
+
+      Set<String> connectedSources = nodeSources.get(node);
+      // if this node is a connector, replace all sources for this node with itself, since a connector is a source
+      if (connectors.contains(node)) {
+        connectedSources = new HashSet<>();
+        connectedSources.add(node);
+        nodeSources.replaceValues(node, connectedSources);
+      }
+      // if more than one source is connected to this node, then this node is a connector.
```

Comment:
I am now really confused. The first if tests whether this node is a connector. Then here we test again, in a different way, whether it is a connector. or do you mean in this comment, "... then this node needs to have a connector preceding it"?